### PR TITLE
Add EnumArray

### PR DESCRIPTION
**Pull request by dangleptr**
It will save lots of space occupied when the Enum has only few values.

### CODE REVIEW

File: src/common/base/EnumArray.h [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        };
+
+        typedef Stat<number, count> Next;
+    };
+
+    template<size_t n>
+    class isPowerTwo {
+    public:
+        enum {
+            ret = (n != 0) && !(n & (n - 1)),
```
Comment (wadeliuyi) [inline src/common/base/EnumArray.h]: this mean ret is true just when n=1,2,4,8,16... like 0x100000...0000?

File: src/common/base/EnumArray.h [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    ~EnumArray() = default;
+
+    void put(size_t idx, E e) {
+        auto v = Transformer::toValue(e);
+        DCHECK(idx < kTotalBits) << "idx " << idx << ", total " << kTotalBits;
+        auto bIdx = blockIndex(idx);
+        auto offset = bitOffset(idx);
+        uint64_t mask = v << offset;
+        blocks_[bIdx] |= mask;
```
Comment (wadeliuyi) [inline src/common/base/EnumArray.h]: is need set 0 to blocks_[bIdx]? if call put with same idx but diffirent e, then maybe can not get correct e?

File: src/common/base/EnumArray.h (hunk 1 above)
Comment (dangleptr) [reply src/common/base/EnumArray.h]: Ha,  it just handles the unsigned integer.

File: src/common/base/EnumArray.h (hunk 2 above)
Comment (dangleptr) [reply src/common/base/EnumArray.h]: Good point!  We need to handle it.

**Pull request closed without merging**